src/models/position_inputs.cpp

```diff
@@ -458,4 +457,149 @@ void PositionInputs::RewindMask(size_t index) {
+    }
+    window_size_ = model_.config_->model.decoder.sliding_window->window_size;
+  }
+
+  if (has_posid_input_) {
+    position_ids_type_ = model_.session_info_->GetInputDataType(model_.config_->model.decoder.inputs.position_ids);
+    if (position_ids_type_ != Ort::TypeToTensorType<int32_t>)
+      throw std::runtime_error("WindowedPositionInputs only supports int32_t position_ids");
+
+    position_ids_shape_ = {1, model_.config_->model.decoder.sliding_window->window_size};
+  }
+
+  if (has_mask_input_) {
+    attention_mask_type_ = model_.session_info_->GetInputDataType(model_.config_->model.decoder.inputs.attention_mask);
+    if (attention_mask_type_ != Ort::TypeToTensorType<int32_t>)
+      throw std::runtime_error("WindowedPositionInputs only supports int32_t attention_mask");
+
+    attention_mask_shape_ = {1, model_.config_->model.context_length};
+  }
+}
+
+void WindowedPositionInputs::Add() {
+  if (has_posid_input_) {
+    position_ids_index_ = state_.inputs_.size();
+    state_.input_names_.push_back(model_.config_->model.decoder.inputs.position_ids.c_str());
+    state_.inputs_.push_back(position_ids_.get());
+  }
+
+  if (has_mask_input_) {
+    attention_mask_index_ = state_.inputs_.size();
+    state_.input_names_.push_back(model_.config_->model.decoder.inputs.attention_mask.c_str());
+    state_.inputs_.push_back(attention_mask_.get());
+  }
+}
+
+void WindowedPositionInputs::Update(DeviceSpan<int32_t> next_tokens, int total_length, int new_length) {
+  if (window_index_ == 0) {
+    num_windows_ = (next_tokens.size() + window_size_ - 1) / window_size_;
+    if (has_posid_input_) {
+      position_ids_ = OrtValue::CreateTensor(model_.allocator_cpu_, position_ids_shape_, position_ids_type_);
+
+      // next_tokens will always be padded so that it's size is a multiple of window_size_
+      // next_tokens -> [0, a, b, c, d, e]
+      // window_size = 3, num_windows = 2, pad_token = 0
+      // window_index = 0, position_ids_ -> [0, 0, 1]
+      auto* position_ids_data = position_ids_->GetTensorMutableData<int32_t>();
+      for (int i = 0, j = 0; i < position_ids_shape_[1]; i++) {
+        if (next_tokens.Span()[i] == model_.config_->model.pad_token_id) {
+          position_ids_data[i] = 0;
+        } else {
+          position_ids_data[i] = j++;
+        }
+      }
+    }
+
+    if (has_mask_input_) {
+      attention_mask_ = OrtValue::CreateTensor(model_.allocator_cpu_, attention_mask_shape_, attention_mask_type_);
+
+      // next_tokens will always be padded so that it's size is a multiple of window_size_
+      // next_tokens -> [0, a, b, c, d, e]
+      // window_size = 3, num_windows = 2, pad_token = 0
+      // window_index = 0, attention_mask_ -> ([0] * context_length - window_size_) + [0, 1, 1]
+      auto* attention_mask_data = attention_mask_->GetTensorMutableData<int32_t>();
+      std::fill(attention_mask_data, attention_mask_data + attention_mask_shape_[1], 0);
```

Comment:
nit: does this only need to fill `attention_mask_shape_[1] - window_size_` zeros? it looks like the next for loop will handle the rest.